File: Exercises/Python/machine-learning-ex4/randInitializeWeights.py
import scipy.io #Used to load the OCTAVE *.mat files
from scipy import linalg, sparse #for linalg
import numpy as np #for lin alg operations
from numpy import random as rand
import itertools

input_layer_size  = 400
hidden_layer_size = 25
num_labels = 10

def randInitializeWeights(L_in, L_out):  
	W=np.zeros((L_out, 1+L_in))
	epsilon_init=.12;
	return np.mat(np.random.rand(L_out, 1+L_in)*2*epsilon_init-epsilon_init);

def flattenParams(thetas_list):
    flattened_list = [ mytheta.flatten() for mytheta in thetas_list ]
    combined = list(itertools.chain.from_iterable(flattened_list))
    #assert len(combined) == (input_layer_size+1)*hidden_layer_size + \
                            #(hidden_layer_size+1)*num_labels
    return np.array(combined).reshape((len(combined),1))

initial_Theta1 = randInitializeWeights(input_layer_size, hidden_layer_size)
initial_Theta2 = randInitializeWeights(hidden_layer_size, num_labels)

theta_list = [ initial_Theta1, initial_Theta2 ]

initial_nn_params = flattenParams(theta_list)

lst  = nnCostFunction(nn_params, input_layer_size, hidden_layer_size, num_labels, X, y, 3);

debug_J=lst['Cost']
grad=lst['Gradient']

print('Cost at (fixed) debugging paramters (this value should be about 0.576051)' + str(debug_J))

File: Exercises/R/machine-learning-ex4/randInitializeWeights.R
library(R.matlab)
library(sigmoid)

pathname_data<-"/Users/cmr/Desktop/Machine Learning/Class/R/ex4/ex4data1.mat"
data <- readMat(pathname_data)

pathname_weights<-"/Users/cmr/Desktop/Machine Learning/Class/R/ex4/ex4weights.mat"
weights <- readMat(pathname_weights)

input_layer_size  <- 400;  # 20x20 Input Images of Digits
hidden_layer_size <- 25;   # 25 hidden units
num_labels <- 10;          # 10 labels, from 1 to 10  

X <- data$X
y <- data$y

randInitializeWeights <- function(L_in, L_out){	
	W<-matrix(0, L_out, 1+L_in)
	epsilon<-.12
	return (matrix(runif(L_out*(L_in+1)), L_out, L_in+1)); #returns matrix that is L_out x (L_in+1)
}

initial_Theta1 <- randInitializeWeights(input_layer_size, hidden_layer_size)
initial_Theta2 <- randInitializeWeights(hidden_layer_size, num_labels)


unrolled_theta1 <- as.matrix(as.vector(Theta1))
unrolled_theta2 <- as.matrix(as.vector(Theta2))

initial_nn_params <- as.matrix(c(unrolled_theta1, unrolled_theta2))

print("\nChecking Backpropagation (w/ Regularization) ... \n")

# Also output the costFunction debugging values
returnedList  <- nnCostFunction(nn_params, input_layer_size, hidden_layer_size, num_labels, X, y, 3);

J<-returnedList$cost
grad<-returnedList$theta

cat("Cost at (fixed) debugging parameters (for lambda = 3, this value should be about 0.576051)", J);







